oraclesrv/keras_model.py

```diff
@@ -0,0 +1,156 @@
+import os, sys
+import traceback
+import time
+import argparse
+
+import pandas as pd
+import numpy as np
+from tensorflow import keras
+from tensorflow.keras import layers
+from sklearn.model_selection import train_test_split
+
+from flask import current_app
+
+try:
+    import cPickle as pickle
+except ImportError:
+    import pickle
+
+
+class KerasModel(object):
+
+    training_file = '/keras_model_files/data.csv'
+
+    units = [16, 16, 8, 8, 1]
+    optimizer = 'adam'
+    loss = 'binary_crossentropy'
+    epoch = 100
+    batch_size = 1
+
+    model_file = '/keras_model_files/4layer4dim.pkl'
+
+    def train(self):
+        """
+
+        :return:
+        """
+
+        df = pd.read_csv(os.path.dirname(__file__) + self.training_file)
+        properties = list(df.columns.values)
+        properties.remove('label')
+        X = df[properties]
+        y = df['label']
+
+        X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)
+        len_test_valid = len(y_test)
+        x_val = X_train[:len_test_valid]
+        partial_x_train = X_train[len_test_valid:]
+        y_val = y_train[:len_test_valid]
+        partial_y_train = y_train[len_test_valid:]
+
+        width = X_train.shape[1]
+        self.model = keras.Sequential([
+            layers.Flatten(input_shape=(width,)),
+            layers.Dense(self.units[0], activation="relu"),
+            layers.Dense(self.units[1], activation="relu"),
+            layers.Dense(self.units[2], activation="relu"),
+            layers.Dense(self.units[3], activation="relu"),
+            layers.Dense(1, activation="sigmoid"),
+        ])
+
+        self.model.compile(optimizer=self.optimizer,
+                      loss=self.loss,
+                      metrics=['accuracy'])
+
+        self.history = self.model.fit(partial_x_train,
+                            partial_y_train,
+                            epochs=self.epoch,
+                            batch_size=self.batch_size,
+                            validation_data=(x_val, y_val),
+                            verbose=0)
+
+        self.test_loss, self.test_accuracy = self.model.evaluate(X_test, y_test)
+
+        return (self.test_accuracy >= 0.95)
+
+    def predict(self, scores):
+        """
+
+        :param scores: list of scores for [abstract, title, author, year]
+        :return:
+        """
+        return float('%.7g' % (np.float32(np.take(self.model.predict([scores]), 0)).item()))
+
+    def save(self):
+        """
+        save object to a pickle file
+
+        """
+        try:
+            with open(os.path.dirname(__file__) + self.model_file, "wb") as f:
```

Comment:
i think it would be better to save the model outside of the python package; using a pre-configured location. Benefits is that you could test/swap different models